swarm/src/dial_opts.rs

```diff
@@ -0,0 +1,205 @@
+    ///   DialOpts::peer_id(PeerId::random())
+    ///      .condition(PeerCondition::Disconnected)
+    ///      .addresses(vec!["/ip6/::1/tcp/12345".parse().unwrap()])
+    ///      .extend_addresses_through_behaviour()
+    ///      .build();
+    ///   ```
+    pub fn peer_id(peer_id: PeerId) -> WithPeerId {
+        WithPeerId {
+            peer_id,
+            condition: Default::default(),
+        }
+    }
+
+    /// Dial an unknown peer.
+    ///
+    ///   ```
+    ///   # use libp2p_swarm::dial_opts::DialOpts;
+    ///   DialOpts::unknown_peer_id()
+    ///      .address("/ip6/::1/tcp/12345".parse().unwrap())
+    ///      .build();
+    ///   ```
+    pub fn unknown_peer_id() -> WithoutPeerId {
+        WithoutPeerId {}
+    }
+
+    /// Get the [`PeerId`] specified in a [`DialOpts`] if any.
+    pub fn get_peer_id(&self) -> Option<PeerId> {
+        match self {
+            DialOpts(Opts::WithPeerId(WithPeerId { peer_id, .. })) => Some(*peer_id),
+            DialOpts(Opts::WithPeerIdWithAddresses(WithPeerIdWithAddresses {
+                peer_id, ..
+            })) => Some(*peer_id),
+            DialOpts(Opts::WithoutPeerIdWithAddress(_)) => None,
+        }
+    }
+}
+
+/// Internal options type.
+///
+/// Not to be constructed manually. Use either of the below instead:
+///
+/// - [`DialOpts::peer_id`] dialing a known peer
+/// - [`DialOpts::unknown_peer_id`] dialing an unknown peer
+#[derive(Debug)]
+pub(super) enum Opts {
+    WithPeerId(WithPeerId),
+    WithPeerIdWithAddresses(WithPeerIdWithAddresses),
+    WithoutPeerIdWithAddress(WithoutPeerIdWithAddress),
+}
+
+#[derive(Debug)]
+pub struct WithPeerId {
+    pub(crate) peer_id: PeerId,
+    pub(crate) condition: PeerCondition,
+}
+
+impl WithPeerId {
+    /// Specify a [`PeerCondition`] for the dial.
+    pub fn condition(mut self, condition: PeerCondition) -> Self {
+        self.condition = condition;
+        self
+    }
+
+    /// Specify a set of addresses to be used to dial the known peer.
+    pub fn addresses(self, addresses: Vec<Multiaddr>) -> WithPeerIdWithAddresses {
+        WithPeerIdWithAddresses {
+            peer_id: self.peer_id,
+            condition: self.condition,
+            addresses,
+            extend_addresses_through_behaviour: false,
+        }
+    }
+
+    /// Build the final [`DialOpts`].
+    ///
+    /// Addresses to dial the peer are retrieved via
+    /// [`NetworkBehaviour::addresses_of_peer`](crate::behaviour::NetworkBehaviour::addresses_of_peer).
+    pub fn build(self) -> DialOpts {
+        DialOpts(Opts::WithPeerId(self))
+    }
+}
+
+#[derive(Debug)]
+pub struct WithPeerIdWithAddresses {
+    pub(crate) peer_id: PeerId,
+    pub(crate) condition: PeerCondition,
+    pub(crate) addresses: Vec<Multiaddr>,
+    pub(crate) extend_addresses_through_behaviour: bool,
+}
+
+impl WithPeerIdWithAddresses {
+    /// Specify a [`PeerCondition`] for the dial.
+    pub fn condition(mut self, condition: PeerCondition) -> Self {
```

Comment:
This API requires users to import one more type: `PeerCondition`. Might be worth providing 3 functions instead that configure the corresponding condition.

Please excuse if that has been discussed and dismissed already!

Reply:
Modelling 3 exclusive states with 3 methods instead of an `enum` seems not intuitive to me. Is the additional import of `PeerCondition` worth the added complexity?

Reply:
I think there is a misunderstanding. I would still model the states internally with an enum. However, nothing stops anyone from doing this:

```rust
DialOpts::address(addr).condition(...).condition(...)
```

It might give people a hint that the condition is overridden but in theory, it could also be implemented in an additive way.
I don't feel strongly about it so happy to keep the enum :)

Reply:
> I think there is a misunderstanding. I would still model the states internally with an enum.

No misunderstanding :)

>  I don't feel strongly about it so happy to keep the enum :)

Will keep the enum for now. Not 100% happy but can't think of a better alternative. We can still further iterate on it. Merging for now.